java-frontend/src/main/java/org/sonar/java/regex/ast/JavaRegexSource.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.regex.ast;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+import org.sonar.java.AnalyzerMessage;
+import org.sonar.java.AnalyzerMessage.TextSpan;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+
+public class JavaRegexSource implements RegexSource {
+
+  private final String sourceText;
+
+  /**
+   * Maps an index of the regular expression to the TextSpan string literal that starts at the index
+   */
+  private final NavigableMap<Integer, TextSpan> indexToTextSpan;
+
+  public JavaRegexSource(List<LiteralTree> stringLiterals) {
+    StringBuilder sb = new StringBuilder();
+    indexToTextSpan = new TreeMap<>();
+    int index = 0;
+    for (LiteralTree literal : stringLiterals) {
+      String text = getString(literal);
+      sb.append(text);
+      TextSpan literalSpan = AnalyzerMessage.textSpanFor(literal);
+      // Create a text span for the string with the quotes stripped out
+      indexToTextSpan.put(index, new TextSpan(literalSpan.startLine, literalSpan.startCharacter + 1, literalSpan.endLine, literalSpan.endCharacter - 1));
+      index += text.length();
+    }
+    sourceText = sb.toString();
+  }
+
+  @Override
+  public String getSourceText() {
+    return sourceText;
+  }
+
+  public List<TextSpan> textSpansFor(IndexRange range) {
+    List<TextSpan> result = new ArrayList<>();
+    Map.Entry<Integer, TextSpan> startEntry = indexToTextSpan.floorEntry(range.getBeginningOffset());
```

Comment:
I really wonder about using `Map.Entry<>`. Can't we use a dedicated type that would encode the logic of the key/value couple with more meaningful names?

I feel it would radically simplify the readiness of the method.